omnipaxos/src/sequence_paxos/leader.rs

```diff
@@ -205,58 +205,59 @@ where
     fn send_accsync(&mut self, to: NodeId) {
         let my_decided_idx = self.get_decided_idx();
         let PromiseMetaData {
-            n: max_promise_n,
-            accepted_idx: max_accepted_idx,
-            ..
+            n: max_promise_n, ..
         } = &self.leader_state.get_max_promise_meta();
         let PromiseMetaData {
-            n: promise_n,
-            accepted_idx: promise_accepted_idx,
+            n: followers_promise_n,
+            accepted_idx: followers_accepted_idx,
             pid,
             ..
         } = self.leader_state.get_promise_meta(to);
-        let follower_decided_idx = self
+        let followers_decided_idx = self
             .leader_state
             .get_decided_idx(*pid)
             .expect("Received PromiseMetaData but not found in ld");
-        let (delta_snapshot, suffix, sync_idx) =
-            if (promise_n == max_promise_n) && (promise_accepted_idx < max_accepted_idx) {
-                if self.internal_storage.get_compacted_idx() > *promise_accepted_idx
-                    && T::Snapshot::use_snapshots()
-                {
-                    let delta_snapshot = self
-                        .internal_storage
-                        .create_diff_snapshot(follower_decided_idx, my_decided_idx)
-                        .expect("storage error while trying to read diff snapshot");
-                    let suffix = self
-                        .internal_storage
-                        .get_suffix(my_decided_idx)
-                        .expect("storage error while trying to read log suffix");
-                    (Some(delta_snapshot), suffix, follower_decided_idx)
-                } else {
-                    let sfx = self
-                        .internal_storage
-                        .get_suffix(*promise_accepted_idx)
-                        .expect("storage error while trying to read log suffix");
-                    (None, sfx, *promise_accepted_idx)
-                }
-            } else if follower_decided_idx < my_decided_idx && T::Snapshot::use_snapshots() {
-                let delta_snapshot = self
+        let (delta_snapshot, suffix, sync_idx) = if (followers_promise_n == max_promise_n) || (*followers_promise_n == self.leader_state.n_leader) {
+            // Follower's accepted entries are valid, send any new entries after
+            if T::Snapshot::use_snapshots() && *followers_accepted_idx < my_decided_idx {
```

Comment:
Is the second check really needed? Because we anyway send the diff snapshots from the follower's decided index.

```suggestion
            if T::Snapshot::use_snapshots() {
```

Reply:
It is not needed for safety but is an optimization (but we need to use `max_accepted_idx` not `followers_accepted_idx` as you've pointed out). The idea is that if the follower has valid accepted entries >= to the leader's `decided_idx` then we don't have to send them a snapshot of the decided entries. We just need to move the follower's `decided_idx` and potentially send a suffix of new accepted entries, and this is exactly what will happen in the `else` statement.

Reply:
Right, makes sense.